### DISCUSSION TITLE
Does someone have a working version of Klarna?

### DISCUSSION THREAD

**Discussion by rinorzek** [swellstores/community#379] [Help] [answered]
Title. I'm struggling for days to get Klarna working with swell, and it's very hard.
Here is what I had at first:
```
export const createKlarnaPayment = async (
  amount: number,
  items: KlarnaOrderItem[],
  owner: KlarnaOwner,
  klarna: KlarnaSourceData,
) => {
  const stripe = await getStripe();
  const returnUrl = getBaseUrl();
  returnUrl.pathname = "/order/submitting";

  const intent = await stripe.createSource({
    type: "klarna",
    flow: "redirect",
    amount,
    currency: "USD",
    owner,
    klarna: {
      product: "payment",
      purchase_country: "US",
      ...klarna,
    },
    source_order: {
      items,
    },
    redirect: {
      return_url: returnUrl.toString(),
    },
  });

  return intent;
};
```
That's all I had in my code, I called this on button click, however I was getting an error on a new page that said:
```
{
  "error": {
    "code": "failed_redirect_url_computation",
    "type": "card_error",
    "message": "Redirect URL computation failed."
  }
}
```
it was no way I could debug this, even after reaching out to swell and stripe support. 

So now I decided to re-implement Klarna through Stripe with their new method of Payment Intent, like shown here. https://stripe.com/docs/payments/klarna/accept-a-payment?platform=web&ui=API

The payment is successful I can assure you that, I check with the Stripe team, but now the problem for sure is Swell in this part, where the order says "Unpaid", what does Swell need in the cart update call with a payment method? The only way I could make the order go in (even though the payment status says "unpaid") was by removing the "method": "klarna" property from the swell.cart.update billing object. If I left that added it complained about this error:
`Payment authorization failed (missing klarna token (billing.klarna.token))`, I don't know what value to get to this property, this is all the params I get from Klaran in the url:
`?payment_intent=pi_3Nrew8He7BhUgpBH0Ivbty4E&payment_intent_client_secret=pi_3Nrew8He7BhUgpBH0Ivbty4E_secret_8fOOxpOaDCXktGvLHFAA20kqj&redirect_status=succeeded#EhcKFWFjY3RfMU1BZDFWSGU3QmhVZ3BCSA&auth=94b377f5-85cf-5d8b-8969-e5edccdd0d23`

**Comment by rinorzek** [comment]
@papertokyo can you help me with this, it's really urgent for me, thanks a lot.

**Comment by rinorzek** [comment]
even after adding the value for token, I used the payment_intent param from the url, I now get a similar error regarding the exp_month property:
`Error: Payment authorization failed (billing.klarna.exp_month missing)` which despite adding it, it still complains.

**Comment by rinorzek** [comment]
<img width="450" alt="Screen Shot 2023-09-18 at 12 57 06" src="https://github.com/swellstores/community/assets/78733927/8856df18-78e4-4368-9554-3e1f6ee70c04">

**Comment by rinorzek** [comment]
how I'm updating the cart:
```
 const cart = await updateCart({
          billing: {
            method: "klarna",
            klarna: {
              exp_month: "12",
              token: paymentIntent,
            },
          },
        });
```

**Comment by rinorzek** [comment]
it's crazy that we can have all of these changes, https://github.com/swellstores/swell-js/commit/48c6d802cda5fae8cfb34661be050654b3249c66 and not a single change in the docs!

**Comment by rinorzek** [comment, marked as answer]
Finally managed to do this by literally having to read the code base:
```
const cart = await swell.cart.update({
          billing: {
            method: "klarna",
            klarna: {
              token: paymentIntent.payment_method,
            },
            intent: {
              stripe: {
                id: paymentIntent.id,
              },
            },
          },
        });
```

Now I have an error saying the account_id is required which is completely meaningless

**Reply by vstudenichnik-insoft** [reply]
Hey @rinorzek, this is the correct cart billing configuration for Klarna payments via Stripe Intent API. You also need to assign an account to the cart in order to convert the cart into an order (`swell.account.login(email, password)`).

**Reply by coderreco** [reply]
Hey @rinorzek, in this instance I assume you also can have guest checkout, but you likely need a sufficient amount of customer information in the billing object as well. I imagine probably first name, last name, and email? [Cart Model](https://developers.swell.is/backend-api/carts/the-cart-model)
Might already be there, but I know I ran into a similar hurdle. Once I added enough billing info it auto-generated an email account id or referenced existing one with a matching email.

**Comment by papertokyo** [comment]
The first thing I'm seeing is you have this code for constructing your return URL:
```
const returnUrl = getBaseUrl();
returnUrl.pathname = "/order/submitting";

...

redirect: {
    return_url: returnUrl.toString(),
},
```
Unless `toString()` is a custom method you have implemented in your `getBaseUrl` function, it seems like it would result in this output which Stripe won't be able to parse, and I would expect to return an error like you're seeing.

![image](https://github.com/swellstores/community/assets/8597335/ebc0db7c-1c77-4016-b116-6fdb5ba94c39)

I will double-check the examples in the docs, but can you verify that the `return_url` value you're passing in the `stripe.createSource` call from your initial post is a valid URL string?
